File: src/test/kotlin/inaction/chap4/innernested/java/ButtonTest.java
package inaction.chap4.innernested.java;

import org.junit.jupiter.api.Test;

import java.io.ByteArrayOutputStream;
import java.io.NotSerializableException;
import java.io.ObjectOutputStream;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertThrows;

class ButtonTest {

    @Test // 자바 ButtonState 가 static 이 아니면 성공
    void javaButtonStateSerializeTest() {
        Button button = new Button();
        Button.ButtonState state = (Button.ButtonState) button.getCurrentState();
        assertThrows(NotSerializableException.class, () -> {
            // 객체를 직렬화하여 ByteArrayOutputStream 에 쓰기 시도
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            ObjectOutputStream oos = new ObjectOutputStream(baos);
            oos.writeObject(state); // 이 시점에서 NotSerializableException 발생 예상
            oos.close();
        });
    }

    @Test // 자바 ButtonState 가 static 이면 성공
    void javaButtonStateSerializeTest2() {
        Button button = new Button();
        Button.ButtonState state = (Button.ButtonState) button.getCurrentState();
        assertDoesNotThrow(() -> {
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            ObjectOutputStream oos = new ObjectOutputStream(baos);
            oos.writeObject(state); // 이 시점에서 NotSerializableException 발생 예상
            oos.close();
        });
    }


}

File: src/test/kotlin/inaction/chap4/innernested/kotlin/ButtonTest.kt
package inaction.chap4.innernested.kotlin

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertDoesNotThrow
import org.junit.jupiter.api.assertThrows
import java.io.ByteArrayOutputStream
import java.io.NotSerializableException
import java.io.ObjectOutputStream

class ButtonTest {

    @Test // 코틀린 ButtonState 가 일반 class 이면 성공
    fun kotlinButtonStateSerializeTest() {
        val button = Button()
        val state: Button.ButtonState = button.getCurrentState() as Button.ButtonState
        assertDoesNotThrow {
            val baos = ByteArrayOutputStream();
            val oos = ObjectOutputStream(baos)
            oos.writeObject(state)
            oos.close()
        }
    }

    @Test // 코틀린 ButtonState 가 inner class 이면 성공
    fun kotlinButtonStateSerializeTest2() {
        val button = Button()
        val state: Button.ButtonState = button.getCurrentState() as Button.ButtonState
        assertThrows<NotSerializableException> {
            val baos = ByteArrayOutputStream();
            val oos = ObjectOutputStream(baos)
            oos.writeObject(state)
            oos.close()
        }
    }
}